### PR TITLE
perf: faster addMessagesSorted

### LINKED ISSUE #465: Improve performance of channelState.addMessageSorted
image

gz#6578

### PR DESCRIPTION
**Pull request by mahboubii**
fixes #465

image

### CODE REVIEW

File: src/channel_state.ts [hunk 1]
```diff
@@ -495,35 +453,50 @@ export class ChannelState<
...
+    if (messages[messages.length - 1].created_at.getTime() < message.created_at.getTime())
+      return messages.concat(message);
+
+    // find the closest index to push the new message
+    let left = 0;
+    let middle = 0;
+    let right = messages.length - 1;
+    while (left <= right) {
+      middle = Math.floor((right + left) / 2);
+      if (messages[middle].created_at.getTime() <= message.created_at.getTime())
```
[the remark this answers is not in the document]
Comment (mahboubii) [reply src/channel_state.ts]: good call

File: src/channel_state.ts [hunk 2]
```diff
@@ -495,35 +453,50 @@ export class ChannelState<
...
+    // if message is newer than last item in the list just concat and return
+    if (messages[messages.length - 1].created_at.getTime() < message.created_at.getTime())
+      return messages.concat(message);
+
+    // find the closest index to push the new message
+    let left = 0;
+    let middle = 0;
+    let right = messages.length - 1;
+    while (left <= right) {
+      middle = Math.floor((right + left) / 2);
```
Comment (nhannah) [inline src/channel_state.ts:470]: In most cases wouldn't a bottom up approach be faster than a binary search given the likelihood of a message being further up is less?

Comment (mahboubii) [reply src/channel_state.ts:470]: L462 takes care of that. cases like pagination could add some 100 more messages to the other end of the array which will be O(n), right now this part is O(log(n)) which is quite fast already

**Review by nhannah — APPROVED** [review]
Looks good lets test

**Review by vishalnarkhede — APPROVED** [review]

File: src/channel.ts [hunk 3]
```diff
@@ -1658,7 +1658,7 @@ export class Channel<
     if (!this.state.messages) {
       this.state.messages = Immutable([]);
     }
-    this.state.addMessagesSorted(messages, true);
+    this.state.addMessagesSorted(messages, false, true);
```
Comment (ferhatelmas) [inline src/channel.ts:1661]: Having options an object with keys would help for readability.

**Review by ferhatelmas — APPROVED** [review]
:100: LGTM

File: src/channel.ts (hunk 3 above)
Comment (mahboubii) [reply src/channel.ts:1661]: 3 params are fine, in case it grows we can change it to object

Comment (ferhatelmas) [reply src/channel.ts:1661]: I agree in general but when they are booleans, it's trickier.

**Pull request merged**